src/shared/cloudformation/templateSymbolResolver.ts

```diff
@@ -0,0 +1,75 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { safeLoad } from 'js-yaml'
+import * as _ from 'lodash'
+import * as vscode from 'vscode'
+import { CloudFormation } from './cloudformation'
+
+export interface TemplateFunctionResource {
+    name: string
+    range: vscode.Range
+}
+
+/**
+ * Provides symbols for a TextDocument.
+ */
+export interface TemplateSymbolProvider {
+    getSymbols(document: vscode.TextDocument): Promise<vscode.DocumentSymbol[]>
+    getText(symbol: vscode.DocumentSymbol, document: vscode.TextDocument): string
+}
+
+/**
+ * Parses symbols in a CloudFormation text document.
+ */
+export class TemplateSymbolResolver {
+    public constructor(
+        private readonly document: vscode.TextDocument,
+        private readonly symbolProvider: TemplateSymbolProvider = new DefaultSymbolProvider()
+    ) {}
+
+    /**
+     * Extracts Function resources from the document.
+     */
+    public async getFunctionResources(): Promise<TemplateFunctionResource[]> {
+        const symbols = await this.symbolProvider.getSymbols(this.document)
+
+        const resources = _(symbols).find({ name: 'Resources', kind: vscode.SymbolKind.Module })?.children || []
+
+        return _(resources)
+            .filter({ kind: vscode.SymbolKind.Module })
+            .filter(resource => this.isFunctionResource(resource))
+            .map(resource => ({ name: resource.name, range: resource.range }))
+            .value()
+    }
+
+    private isFunctionResource(symbol: vscode.DocumentSymbol) {
+        const typeSymbol = _(symbol.children).find({ name: 'Type', kind: vscode.SymbolKind.String })
+
+        if (!typeSymbol) {
+            return false
+        }
+
+        // tslint:disable-next-line: no-unsafe-any
+        const typeValue: string | undefined = safeLoad(this.symbolProvider.getText(typeSymbol, this.document)).Type
+
+        return typeValue === CloudFormation.SERVERLESS_FUNCTION_TYPE
+    }
+}
+
+class DefaultSymbolProvider implements TemplateSymbolProvider {
```

Comment:
similar comment about abstraction. why not let callers decide what they want via function parameters, that is a nicer API than lots of extra classes. And we may not ever _need_ the flexibility, so the classes are pure cost until then.

Reply:
This was done to remove dependencies on `vscode.commands.executeCommand()` and `document.getText()` during testing.

I think this would be less necessary if VSCode offered fakes for these things that we could inject in our test environment. But I'm not aware of this.

As it stands, an optional "provider" parameter seemed like a good solution. Normal users of the class don't need to provide this. I was looking for something similar to java where you can specify a package-private alternate constructor to inject a different impl. Thoughts?

Reply:
> I think this would be less necessary if VSCode offered fakes for these things that we could inject in our test environment.

In some of my recent changes, I've started using sinon to stub out parts of the vs code api that are hit during tests. The downside with that approach is having to know what needs stubbing, and putting those stubs in on a per-test basis.

Based on the existing local debug related code, fewer abstractions will help keep the code easier to understand. This change seems pretty self contained, so I'm neutral on this one.

Reply:
why do we care about stubbing vscode API? It is available to tests so that it can be used, and so that tests actually test real behavior.

Reply:
> fewer abstractions will help keep the code easier to understand.

Agree but my fear is that fewer abstractions can make things harder to test. And making things harder to test usually means the developer will not write tests to cover them. Or if they do, the tests are really complicated / fragile and require a whole bunch of state to be setup before they run.

>  It is available to tests so that it can be used, and so that tests actually test real behavior.

There was a discussion in the past about the desire to run unit tests without the need for an extension host. This probably ties back to the discussion about the level of stubbing done in unit tests versus functional tests.